### ISSUE TITLE
FlashAttentionScore算子开箱性能很差

### ISSUE THREAD

**Issue by CYa9-a11y** [triton-lang/triton-ascend#271]
### Describe the issue

@triton.autotune(

     configs=[
           triton.Config({'BQ': 256, 'BK': 128}, num_warps=8),
           triton.Config({'BQ': 128, 'BK': 64}, num_warps=8),
           triton.Config({'BQ': 64, 'BK': 64}, num_warps=8),
           triton.Config({'BQ': 64, 'BK': 32}, num_warps=4),
           triton.Config({'BQ': 32, 'BK': 64}, num_warps=4),
           triton.Config({'BQ': 32, 'BK': 32}, num_warps=4),
           triton.Config({'BQ': 16, 'BK': 16}, num_warps=2),
       ],
       key=['n', 'D'],  # 当 n 或 D 变化时重新调优
  )
  @triton.jit
  def flash_attention_forward_kernel(
       q,
       k,
       v,
       out,
       l,
       stride_qb,
       stride_qn,
       stride_qd,
       stride_kb,
       stride_kn,
       stride_kd,
       stride_vb,
       stride_vn,
       stride_vd,
       stride_ob,
       stride_on,
       stride_od,
       stride_lb,
       stride_ln,
       n: tl.int32,
       d_scale: tl.float32,
       IS_CAUSAL: tl.constexpr,
       BQ: tl.constexpr,
       BK: tl.constexpr,
       D: tl.constexpr,
       eps: tl.constexpr,
   ):
     pid_b = tl.program_id(0)
    pid_tq = tl.program_id(1)

    q_block_ptr = tl.make_block_ptr(
         base=q + pid_b * stride_qb,
         shape=(n, D),
         strides=(stride_qn, stride_qd),
         offsets=(pid_tq * BQ, 0),
        block_shape=(BQ, D),
         order=(1, 0),
     )
     k_block_ptr = tl.make_block_ptr(
         base=k + pid_b * stride_kb,
         shape=(D, n),
         strides=(stride_kd, stride_kn),
         offsets=(0, 0),
        block_shape=(D, BK),
         order=(0, 1),
     )
     v_block_ptr = tl.make_block_ptr(
         base=v + pid_b * stride_vb,
         shape=(n, D),
        strides=(stride_vn, stride_vd),
         offsets=(0, 0),
        block_shape=(BK, D),
         order=(1, 0),
     )
     o_block_ptr = tl.make_block_ptr(
         base=out + pid_b * stride_ob,
         shape=(n, D),
         strides=(stride_on, stride_od),
         offsets=(pid_tq * BQ, 0),
         block_shape=(BQ, D),
         order=(1, 0),
     )
     l_ptrs = l + pid_b * stride_lb + (pid_tq * BQ + tl.arange(0, BQ))

     m_i = tl.full([BQ], value=float("-inf"), dtype=tl.float32)
     l_i = tl.zeros([BQ], dtype=tl.float32)
     o_i = tl.zeros([BQ, D], dtype=tl.float32)

     q_i = tl.load(q_block_ptr, boundary_check=(0, 1))
     q_i *= d_scale

     j = 0

     loop_end = tl.cdiv(n, BK)
     if IS_CAUSAL:
         loop_end = tl.cdiv((pid_tq + 1) * BQ, BK)

     for j in range(loop_end):
         k_j = tl.load(k_block_ptr, boundary_check=(0, 1))
         v_j = tl.load(v_block_ptr, boundary_check=(0, 1))

         s_ij = tl.dot(q_i, k_j)

         if IS_CAUSAL:
             offs_q = pid_tq * BQ + tl.arange(0, BQ)
             offs_k = j * BK + tl.arange(0, BK)
             causal_mask = offs_q[:, None] >= offs_k[None, :]
             s_ij = tl.where(causal_mask, s_ij, -1e9)

         m_new = tl.maximum(m_i, tl.max(s_ij, axis=1))
         scale = tl.exp(m_i - m_new)
         p_ij = tl.exp(s_ij - m_new[:, None])

         l_new = scale * l_i + tl.sum(p_ij, axis=1)
         o_i = scale[:, None] * o_i + tl.dot(p_ij.to(v_j.dtype), v_j)

         l_i = l_new
         m_i = m_new

         k_block_ptr = tl.advance(k_block_ptr, (0, BK))
         v_block_ptr = tl.advance(v_block_ptr, (BK, 0))

     o_i /= l_i[:, None]
     l_i = m_i + tl.log(l_i + eps)

     tl.store(o_block_ptr, o_i.to(out.dtype.element_ty), boundary_check=(0, 1))
     tl.store(l_ptrs, l_i, mask=(pid_tq * BQ + tl.arange(0, BQ)) < n)

 class FlashAttention(torch.autograd.Function):
     @staticmethod
     def forward(
         ctx,
         q: Float[Tensor, "b n d"],
         k: Float[Tensor, "b n d"],
         v: Float[Tensor, "b n d"],
         is_causal=False,
         eps=1e-6,
     ):
         device = q.device
         b, n, d = q.shape
         scale = 1.0 / math.sqrt(d)

         out = torch.empty_like(q)
         l = torch.empty((b, n), dtype=torch.float32, device=device)

         grid = (b, triton.cdiv(n, 16))

         flash_attention_forward_kernel[grid](
             q,
             k,
             v,
             out,
             l,
             q.stride(0),
             q.stride(1),
             q.stride(2),
             k.stride(0),
             k.stride(1),
            k.stride(2),
            v.stride(0),
             v.stride(1),
             v.stride(2),
             out.stride(0),
             out.stride(1),
             out.stride(2),
             l.stride(0),
             l.stride(1),
             d_scale=scale,
             IS_CAUSAL=is_causal,
             D=d,
             eps=eps,
         )
         ctx.save_for_backward(q, k, v, l, out)
         ctx.is_causal = is_causal
         return out

### Environment details

950上和910B4上性能都特别差，如batch, seq_len, head_dim = 2, 2048, 256时，性能和torch.nn.functional.scaled_dot_product_attention差十倍以上
950上用的是CANN9.0Triton ascend 3.2.1

**Comment by CYa9-a11y** [comment]
测试脚本类似以下这种，采取了多次预热运行取平均

**Comment by CYa9-a11y** [comment]
device = 'npu:0'
torch.manual_seed(42)

batch, seq_len, head_dim = 1, 4096, 512
causal = False

q = torch.randn(batch, seq_len, head_dim, dtype=torch.float32, device=device)
k = torch.randn(batch, seq_len, head_dim, dtype=torch.float32, device=device)
v = torch.randn(batch, seq_len, head_dim, dtype=torch.float32, device=device)

import torch_npu

for _ in range(5):
    _ = FlashAttention.apply(q, k, v, causal, 1e-6)
torch.npu.synchronize()

for _ in range(10):
    _ = FlashAttention.apply(q, k, v, causal, 1e-6)
torch.npu.synchronize()

**Comment by WuTYSFG** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.
